server/entitymanager/entitymanager.go

```diff
@@ -2,38 +2,162 @@
-	return nil, status.Errorf(codes.Unimplemented, "Unimplemented")
+	if e, ok := m.chassisInventory[*lookup]; ok {
+		return e, nil
+	}
+
+	return nil, status.Errorf(codes.NotFound, "chassis %+v not found in inventory", *lookup)
+}
+
+func (m *InMemoryEntityManager) GetBootstrapData(c *bootz.ControlCard) (*bootz.BootstrapDataResponse, error) {
+	// First check if we are expecting this control card.
+	if c.SerialNumber == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "no serial number provided")
+	}
+	if _, ok := m.controlCardStatuses[c.GetSerialNumber()]; !ok {
+		return nil, status.Errorf(codes.NotFound, "control card %v not found in inventory", c.GetSerialNumber())
+	}
+	// Construct the response. This emulator hardcodes these values but a real Bootz server would not.
+	// TODO: Populate these placeholders with realistic ones.
+	return &bootz.BootstrapDataResponse{
+		SerialNum: c.SerialNumber,
+		IntendedImage: &bootz.SoftwareImage{
+			Name:          "Default Image",
+			Version:       "1.0",
+			Url:           "https://path/to/image",
+			OsImageHash:   "ABCDEF",
+			HashAlgorithm: "SHA256",
+		},
+		BootPasswordHash: "ABCD123",
+		ServerTrustCert:  "FakeTLSCert",
+		BootConfig: &bootz.BootConfig{
+			VendorConfig: []byte("Vendor Config"),
+			OcConfig:     []byte("OC Config"),
+		},
+		Credentials: &bootz.Credentials{},
+		// TODO: Populate pathz, authz and certificates.
+	}, nil
+}
+
+func (m *InMemoryEntityManager) SetStatus(req *bootz.ReportStatusRequest) error {
+	if len(req.GetStates()) == 0 {
+		return status.Errorf(codes.InvalidArgument, "no control card states provided")
+	}
+	log.Infof("Bootstrap Status: %v: Status message: %v", req.GetStatus(), req.GetStatusMessage())
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for _, c := range req.GetStates() {
+		previousStatus, ok := m.controlCardStatuses[c.GetSerialNumber()]
+		if !ok {
+			return status.Errorf(codes.NotFound, "control card %v not found in inventory", c.GetSerialNumber())
+		}
+		log.Infof("control card %v changed status from %v to %v", c.GetSerialNumber(), previousStatus, c.GetStatus())
+		m.controlCardStatuses[c.GetSerialNumber()] = c.GetStatus()
+	}
+	return nil
+}
+
+// Sign unmarshals the SignedResponse bytes then generates a signature from its Ownership Certificate private key.
+func (m *InMemoryEntityManager) Sign(resp *bootz.GetBootstrapDataResponse, serial string) error {
+	// Sign the response
+	block, _ := pem.Decode([]byte(m.artifacts.OC.Key))
+	if block == nil {
+		return status.Errorf(codes.Internal, "unable to decode OC private key")
+	}
+	priv, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err != nil {
+		return err
+	}
+	if resp.GetSignedResponse() == nil {
+		return status.Errorf(codes.InvalidArgument, "empty signed response")
+	}
+	signedResponseBytes, err := proto.Marshal(resp.GetSignedResponse())
+	if err != nil {
+		return err
+	}
+	hashed := sha256.Sum256(signedResponseBytes)
+	sig, err := rsa.SignPKCS1v15(nil, priv, crypto.SHA256, hashed[:])
```

Comment:
I think ECs are becoming more common and so I think it'd be a good idea to somehow detect the priv key type and support both algorithms. could be done as an improvement in a later PR though.